src/string/extglobmatch.c

```diff
@@ -0,0 +1,513 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright © 2014 George Spelvim <linux@horizon.com>
+ * Copyright © 2016 Matthias "ailu" Günzel <a1lu@arcor.de>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "extglobmatch.h"
+#include "criterion/internal/common.h"
+
+struct context{
+    int repeat;
+    int except;
+    int nested;
+    char const *pat;
+    char const *str;
+    char const *pat_chunk_end;
+    char const *back_pat;
+    char const *back_str;
+    char const *repeat_pat;
+    char const *repeat_str;
+};
+
+struct ext_glob{
+    char const *cur_pat;
+    char const *end_pat;
+    int ret;
+};
+
+
+static char const *find_pattern_end(char const *pat) {
+    int len = 0;
+    int paren = 0;
+    int patlen = strlen(pat);
+    do {
+        if (*pat == '(')
+            paren++;
+        else if (*pat == ')')
+            paren--;
+        ++len; ++pat;
+    } while (paren > 0 && len < patlen);
+    if (paren) {
+        fprintf(stderr, "pattern error: mismatching parenthesis\n");
+        exit(3);
+    }
+    return pat;
+}
+
+static struct ext_glob find_next(struct ext_glob con) {
+    if (*con.cur_pat == '(') {
+        con.end_pat = find_pattern_end(con.cur_pat);
+        ++con.cur_pat;
+        return con;
+    }
+
+    int count = 0;
+    unsigned char p = *con.cur_pat++;
+    while (count >= 0 && p != '\0' && con.cur_pat < con.end_pat) {
+        switch(p) {
+            case '(': ++count;
+                      break;
+            case ')': --count;
+                      break;
+            case '|': if (count == 0)
+                          return con;
+                      break;
+        }
+        p = *con.cur_pat++;
+    }
+    con.ret = 1;
+    return con;
+}
+
+CR_NORETURN static void malformed_pattern(char p) {
+    fprintf(stderr, "pattern error: Unexpected '%c'\n", p);
+    exit(3);
+}
+
+char const *repeat_stack[1000];
```

Comment:
Private global variables should be marked static. And for that matter, you should avoid global variables altogether and pass the stack pointer to your functions directly (extmatch should be re-entrant).